File: recursion/python/solutions/zip_with.py
"""Solution for zip_with."""

from linked_list import EMPTY_LIST, Node, is_empty, prepend, unprepend, to_python_list


def zip_with(fn, lst1, lst2):
    """
    Combine two lists using fn, stopping at shorter list.

    zip_with(add, 1 -> 2 -> 3, 10 -> 20) => 11 -> 22
    """
    if is_empty(lst1) or is_empty(lst2):
        return EMPTY_LIST

    first1, rest1 = unprepend(lst1)
    first2, rest2 = unprepend(lst2)

    return prepend(fn(first1, first2), zip_with(fn, rest1, rest2))


def add(a, b):
    return a + b


def multiply(a, b):
    return a * b


if __name__ == '__main__':
    # Test cases
    assert to_python_list(zip_with(add, Node.from_list([1, 2, 3]), Node.from_list([10, 20]))) == [11, 22]
    assert to_python_list(zip_with(add, EMPTY_LIST, Node.from_list([1, 2]))) == []
    assert to_python_list(zip_with(multiply, Node.from_list([1, 2, 3]), Node.from_list([2, 3, 4]))) == [2, 6, 12]

    print("All tests passed!")

File: recursion/javascript/solutions/zipWith.js
/**
 * Solution for zipWith.
 */

const { EMPTY_LIST, Node, isEmpty, prepend, unprepend, toArray } = require('./linkedList');

function zipWith(fn, list1, list2) {
  if (isEmpty(list1) || isEmpty(list2)) {
    return EMPTY_LIST;
  }

  let [first1, rest1] = unprepend(list1);
  let [first2, rest2] = unprepend(list2);

  return prepend(fn(first1, first2), zipWith(fn, rest1, rest2));
}

const add = (a, b) => a + b;
const multiply = (a, b) => a * b;

if (require.main === module) {
  console.assert(JSON.stringify(toArray(zipWith(add, Node.fromArray([1, 2, 3]), Node.fromArray([10, 20])))) === '[11,22]');
  console.assert(JSON.stringify(toArray(zipWith(multiply, Node.fromArray([1, 2, 3]), Node.fromArray([10, 20, 30])))) === '[10,40,90]');
  console.assert(JSON.stringify(toArray(zipWith(add, EMPTY_LIST, Node.fromArray([1, 2])))) === '[]');

  console.log('All tests passed!');
}

module.exports = { zipWith };
